subprojects/core/src/main/java/org/gradle/caching/internal/services/NextGenBuildCacheControllerFactory.java

```diff
@@ -103,10 +104,42 @@ protected BuildCacheController doCreateController(
 
     private static NextGenBuildCacheHandler resolveService(@Nullable DescribedBuildCacheService<? extends BuildCache, ? extends BuildCacheService> describedService) {
         return describedService != null && describedService.config.isEnabled()
-            ? new DefaultNextGenBuildCacheHandler(describedService.service, describedService.config.isPush())
+            ? new DefaultNextGenBuildCacheHandler(makeCompatible(describedService.service), describedService.config.isPush())
             : DISABLED_BUILD_CACHE_HANDLER;
     }
 
+    private static NextGenBuildCacheService makeCompatible(BuildCacheService service) {
+        if (service instanceof NextGenBuildCacheService) {
+            return (NextGenBuildCacheService) service;
+        }
+        return new NextGenBuildCacheService() {
```

Comment:
🤔 Similarly as before, should we rather throw an exception here and we forbid to use old services with the new API?

Reply:
I think we should keep this one. It would be nice to allow people who use current-gen custom backends to try the new cache. If we see it's a problem, we can always revert on this decision.